rolling-shutter/medley/broker.go

```diff
@@ -0,0 +1,184 @@
+// set in the constructor.
+func (b *Broker[T]) Start(initialSubscriptions ...chan T) {
+	subscriptions := map[chan T]struct{}{}
+	for _, sub := range initialSubscriptions {
+		log.Debug().Int("num-subscribers", len(subscriptions)).Msg("subscribed initial subscription")
+		subscriptions[sub] = struct{}{}
+	}
+	started := make(chan struct{})
+	go func() {
+		b.mux.Lock()
+		b.running = true
+		b.mux.Unlock()
+		close(started)
+		for {
+			select {
+			case <-b.stop:
+				for msgCh := range subscriptions {
+					close(msgCh)
+					delete(subscriptions, msgCh)
+				}
+				b.mux.Lock()
+				b.running = false
+				b.mux.Unlock()
+				return
+			case msgCh := <-b.subscribe:
+				subscriptions[msgCh] = struct{}{}
+			case msgCh := <-b.unsubscribe:
+				delete(subscriptions, msgCh)
+				close(msgCh)
+				log.Debug().Int("num-subscribers", len(subscriptions)).Msg("Unsubscribed")
+			case msg := <-b.publish:
+				if b.blockingSend {
+					// use blocking send to make sure all subscribers
+					// got the value,
+					// could block the whole broker (+ successive Publish() calls) if a receiver
+					// is currently busy
+					for msgCh := range subscriptions {
+						msgCh <- msg
+					}
+				} else {
+					// use non-blocking send to protect the broker
+					// and calls to Publish()
+					// receivers could miss published values when they are currently busy
+					// -> higher non-zero 'bufferSizeSubscriber' values could circumvent this somewhat
+					for msgCh := range subscriptions {
+						select {
+						case msgCh <- msg:
+						default:
+						}
+					}
+				}
+			}
+		}
+	}()
+	<-started
+}
+
+// `Stop` will close the internal stop channel
+// and will stop the Brokers internal loop
+// as soon as the closed stop channel is selected.
+// A call to Stop() will also cause the subscribed
+// channels to be closed.
+func (b *Broker[_]) Stop() {
+	close(b.stop)
+}
+
+// `Subscribe` registers the channel to the internal
+// send operation. It will receive a published value
+// as soon as the other registered subscribers in line
+// received the value (in the blockingSend case),
+// or as soon as the value is published AND the subscriber
+// is currently waiting on the receive operation
+// (in the non blockingSend case).
+// This is blocking, even if the loop is not running.
+func (b *Broker[T]) Subscribe() chan T {
+	channel := make(chan T, b.bufferSizeSubscriber)
+	b.subscribe <- channel
+	return channel
+}
+
+// `Unsubscribe` deregisters the subscribed channel
+// from the internal send operation and close the
+// subscribed channel.
+// This is a noop when the loop is not running.
+func (b *Broker[T]) Unsubscribe(channel chan T) {
+	b.mux.RLock()
+	if !b.running {
+		// return directly in case the loop is not running,
+		// otherwise this would block forever
+		b.mux.RUnlock()
+		return
+	}
+	b.mux.RUnlock()
+	for {
+		select {
+		case <-b.stop:
+			return
+		case b.unsubscribe <- channel:
+			return
+		case <-channel:
```

Comment:
I don't understand this, what does this do?

Reply:
So this one tries a receive on the passed in channel, 
while the case above sends the channel itself to the `b.unsubscribe` channel ;)

It is there because the `Unsubscribe()` e.g. could be called from the same goroutine that usually consumes values
from the channel. But in a blockingSend broker, the broker's loop could be in the process of publishing values, thus it depends on all subscribers still being able to consume the values. The noop `<-channel` allows this while waiting for the channel to be unsubscribed.
Which in turn can only happen if the brokers loop is able to finish the publishing and can receive on the `b.unsubscribe` channel. 
Does this makes sense? I know it's a lot ...